### PR TITLE
Create `@huggingface/jinja` for parsing and rendering Jinja chat templates

### PR DESCRIPTION
**Pull request by xenova**
This PR introduces the `@huggingface/jinja` library, which is a minimalistic JavaScript implementation of the Jinja templating engine, specifically designed for parsing ML chat templates.

Although it was [originally created](https://github.com/xenova/transformers.js/pull/408) for (and integrated into) transformers.js, it became clear that others can use this functionality too, without the overhead of the transformers.js library.

**Example usage:** Loading a `tokenizer_config.json` from the HF hub and render a list of messages

```js
import { Template } from "@huggingface/templates";
import { downloadFile } from "@huggingface/hub";

const config = await (await downloadFile({
    repo: "mistralai/Mistral-7B-Instruct-v0.1",
    path: "tokenizer_config.json"
})).json();

const chat = [
    { "role": "user", "content": "Hello, how are you?" },
    { "role": "assistant", "content": "I'm doing great. How can I help you today?" },
    { "role": "user", "content": "I'd like to show off how chat templating works!" },
];

const template = new Template(config.chat_template);
const result = template.render({
    messages: chat,
    bos_token: config.bos_token,
    eos_token: config.eos_token,
});
// "<s>[INST] Hello, how are you? [/INST]I'm doing great. How can I help you today?</s> [INST] I'd like to show off how chat templating works! [/INST]"
```

### TODO:
- [ ] Auto-generate documentation
- [x] Trigger unit tests for GH actions
- [x] Add support for array slicing (e.g., `messages[1:]`)
- [x] Do `trim_blocks=True, lstrip_blocks=True` (needed by [HuggingFaceH4/zephyr-7b-beta](https://huggingface.co/HuggingFaceH4/zephyr-7b-beta))
- [x] Add test cases for invalid templates
- [ ] Library minification
- [ ] Publish to NPM
- [ ] Integration into transformers.js (will happen after this is published) as optional dependency

### CODE REVIEW

**Comment by xenova** [comment]
Regarding the above TODOs, I would appreciate some assistance from core maintainers of the repo to follow best practices (and to be aligned with the other packages) :)

**Comment by xenova** [comment]
Thanks @coyotte508 😇 (I don't know how I missed that guide, sorry about that).

Since the codebase was migrated from transformers.js, it's using JSDoc at the moment, and I think this might cause some issues with the current project structure. Would you recommend I switch over to using typescript? I don't mind either way. If I do switch, it will also probably help with the migration to tsup instead of vite.

**Comment by xenova** [comment]
Okay I've finished rewriting in typescript, and the node tests are passing in GH actions. However, the browser test seems to be timing out after 10 minutes?
```
"test:browser": "vitest run --browser.name=chrome --browser.headless",
```

any ideas @coyotte508 ?

File: pnpm-workspace.yaml [hunk 1]
```diff
@@ -7,3 +7,4 @@ packages:
   - "packages/languages"
   - "packages/tasks"
   - "packages/widgets"
+  - "packages/templates"
```
Comment (julien-c) [inline pnpm-workspace.yaml:10]:
```suggestion
  - "packages/chat_templates"
```

`chat_templates` (or `chat-templates`?) would be clearer no?

File: packages/templates/package.json [hunk 2]
```diff
@@ -0,0 +1,49 @@
+{
+	"name": "@huggingface/templates",
```
Comment (julien-c) [inline packages/templates/package.json:2]:
```suggestion
	"name": "@huggingface/chat-templates",
```

**Review by pcuenca — COMMENTED** [review]
Very nice! I had in mind some manual parsing / regexps for swift-transformers, I'll (eventually) copy this instead.

File: pnpm-workspace.yaml (hunk 1 above)
Comment (xenova) [reply pnpm-workspace.yaml:10]: Sure, I can change it 😇 I originally opted for just "templates" because the library can actually also be used outside of an ML context, but with our focus being quite clear (i.e., minimalistic and specifically designed for ChatML formats), it would probably make more sense to rename it.

Comment (julien-c) [reply pnpm-workspace.yaml:10]: if it's just going to be the current scope, you can also name it `@huggingface/jinja` i guess 

Depends on the planned scope!

**Comment by xenova** [comment]
Removing vitest from my dev dependency list indeed fixed the freezing issue 👍 The browser tests are now failing because it's unable to resolve `hash-wasm`. I assume vite is doing some unnecessary bundling, because we aren't using this module in the tests (we aren't uploading files > 10MB). Fortunately, it looks like there are already plans to remove it ([issue](https://github.com/huggingface/huggingface.js/issues/221)).

```
packages/chat-templates test:browser$ vitest run --browser.name=chrome --browser.headless
packages/chat-templates test:browser:  RUN  v0.29.7 /home/runner/work/huggingface.js/huggingface.js/packages/chat-templates
packages/chat-templates test:browser:       Browser runner started at http://localhost:63315
packages/chat-templates test:browser: The following dependencies are imported but could not be resolved:
packages/chat-templates test:browser:   hash-wasm (imported by /home/runner/work/huggingface.js/huggingface.js/packages/hub/src/utils/sha256.ts)
packages/chat-templates test:browser: Are they installed?
packages/chat-templates test:browser: Failed to resolve import "hash-wasm" from "../hub/src/utils/sha256.ts?v=1700763447278". Does the file exist?
packages/chat-templates test:browser: 6:17:27 PM [vite] Internal server error: Failed to resolve import "hash-wasm" from "../hub/src/utils/sha256.ts?v=1700763447278". Does the file exist?
packages/chat-templates test:browser:   Plugin: vite:import-analysis
packages/chat-templates test:browser:   File: /home/runner/work/huggingface.js/huggingface.js/packages/hub/src/utils/sha256.ts?v=1700763447278:150:29
packages/chat-templates test:browser:   123|      }
packages/chat-templates test:browser:   124|      if (!wasmModule) {
packages/chat-templates test:browser:   125|        wasmModule = await import("hash-wasm");
packages/chat-templates test:browser:      |                                  ^
packages/chat-templates test:browser:   126|      }
packages/chat-templates test:browser:   127|      const sha2562 = await wasmModule.createSHA256();
packages/chat-templates test:browser:       at formatError (file:///home/runner/work/huggingface.js/huggingface.js/node_modules/.pnpm/vite@4.1.4_@types+node@18.[14](https://github.com/huggingface/huggingface.js/actions/runs/6973087542/job/18976473121#step:11:15).6/node_modules/vite/dist/node/chunks/dep-ca21228b.js:41418:46)
packages/chat-templates test:browser:       at TransformContext.error (file:///home/runner/work/huggingface.js/huggingface.js/node_modules/.pnpm/vite@4.1.4_@types+node@18.14.6/node_modules/vite/dist/node/chunks/dep-ca21228b.js:41414:19)
packages/chat-templates test:browser:       at normalizeUrl (file:///home/runner/work/huggingface.js/huggingface.js/node_modules/.pnpm/vite@4.1.4_@types+node@18.14.6/node_modules/vite/dist/node/chunks/dep-ca21228b.js:39706:33)
packages/chat-templates test:browser:       at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
packages/chat-templates test:browser:       at async TransformContext.transform (file:///home/runner/work/huggingface.js/huggingface.js/node_modules/.pnpm/vite@4.1.4_@types+node@18.14.6/node_modules/vite/dist/node/chunks/dep-ca21228b.js:39840:47)
packages/chat-templates test:browser:       at async Object.transform (file:///home/runner/work/huggingface.js/huggingface.js/node_modules/.pnpm/vite@4.1.4_@types+node@18.14.6/node_modules/vite/dist/node/chunks/dep-ca21228b.js:41685:30)
packages/chat-templates test:browser:       at async loadAndTransform (file:///home/runner/work/huggingface.js/huggingface.js/node_modules/.pnpm/vite@4.1.4_@types+node@18.14.6/node_modules/vite/dist/node/chunks/dep-ca21228b.js:39479:29)
packages/chat-templates test:browser:  ❯ test/e2e.test.js  (0 test)
packages/chat-templates test:browser:  ✓ test/templates.test.js  (3 tests) 19ms
packages/chat-templates test:browser: ⎯⎯⎯⎯⎯⎯ Failed Suites 1 ⎯⎯⎯⎯⎯⎯⎯
packages/chat-templates test:browser:  FAIL  test/e2e.test.js [ test/e2e.test.js ]
packages/chat-templates test:browser: TypeError: Failed to fetch dynamically imported module: http://localhost:63315/home/runner/work/huggingface.js/huggingface.js/packages/chat-templates/test/e2e.test.js?v=1700763447278
packages/chat-templates test:browser: ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[1/1]⎯
packages/chat-templates test:browser:  Test Files  1 failed | 1 passed (2)
packages/chat-templates test:browser:       Tests  3 passed (3)
packages/chat-templates test:browser:    Start at  18:[17](https://github.com/huggingface/huggingface.js/actions/runs/6973087542/job/18976473121#step:11:18):13
packages/chat-templates test:browser:    Duration  14.84s (transform 0ms, setup 0ms, collect 60ms, tests [19](https://github.com/huggingface/huggingface.js/actions/runs/6973087542/job/18976473121#step:11:20)ms)
packages/chat-templates test:browser: Failed
```

File: pnpm-workspace.yaml (hunk 1 above)
Comment (julien-c) [reply pnpm-workspace.yaml]: discussing more with @gary149 and others i kinda like `@huggingface/jinja` too – do you think this package will include (or already does include) more stuff than just a jinja parser?

**Comment by xenova** [comment]
> you should change this line to use `@huggingface/hub` then: https://github.com/huggingface/huggingface.js/pull/352/files#diff-a22101bc7201b13f434e24905cd2b71d26dd97e31276eee1229f765f1df0b980R5

I did try that, but I got:
```
 RUN  v0.29.7 D:/joshu/Xenova/Projects/huggingface.js/packages/chat-templates

 ❯ test/e2e.test.js (0)
 ✓ test/integration.test.js (1)
 ✓ test/templates.test.js (3)

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ Failed Suites 1 ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯
 FAIL  test/e2e.test.js [ test/e2e.test.js ]
Error: Failed to resolve entry for package "@huggingface/hub". The package may have incorrect main/module/exports specified in its package.json.
 ❯ packageEntryFailure ../../node_modules/.pnpm/vite@4.1.4_@types+node@18.14.6/node_modules/vite/dist/node/chunks/dep-ca21228b.js:22005:11
 ❯ resolvePackageEntry ../../node_modules/.pnpm/vite@4.1.4_@types+node@18.14.6/node_modules/vite/dist/node/chunks/dep-ca21228b.js:22002:5
 ❯ tryNodeResolve ../../node_modules/.pnpm/vite@4.1.4_@types+node@18.14.6/node_modules/vite/dist/node/chunks/dep-ca21228b.js:21737:20
 ❯ Context.resolveId ../../node_modules/.pnpm/vite@4.1.4_@types+node@18.14.6/node_modules/vite/dist/node/chunks/dep-ca21228b.js:21488:28
 ❯ Object.resolveId ../../node_modules/.pnpm/vite@4.1.4_@types+node@18.14.6/node_modules/vite/dist/node/chunks/dep-ca21228b.js:41612:46
 ❯ TransformContext.resolve ../../node_modules/.pnpm/vite@4.1.4_@types+node@18.14.6/node_modules/vite/dist/node/chunks/dep-ca21228b.js:41363:23
 ❯ normalizeUrl ../../node_modules/.pnpm/vite@4.1.4_@types+node@18.14.6/node_modules/vite/dist/node/chunks/dep-ca21228b.js:39698:34
 ❯ TransformContext.transform ../../node_modules/.pnpm/vite@4.1.4_@types+node@18.14.6/node_modules/vite/dist/node/chunks/dep-ca21228b.js:39840:47        
 ❯ Object.transform ../../node_modules/.pnpm/vite@4.1.4_@types+node@18.14.6/node_modules/vite/dist/node/chunks/dep-ca21228b.js:41685:30
 ❯ loadAndTransform ../../node_modules/.pnpm/vite@4.1.4_@types+node@18.14.6/node_modules/vite/dist/node/chunks/dep-ca21228b.js:39479:29

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[1/1]⎯

 Test Files  1 failed | 2 passed (3)
      Tests  4 passed (4)
   Start at  21:25:55
   Duration  918ms (transform 283ms, setup 0ms, collect 352ms, tests 15ms)

```
👀

File: pnpm-workspace.yaml (hunk 1 above)
Comment (xenova) [reply pnpm-workspace.yaml]: Yep, it's just jinja at the moment, and it probably won't change unless we end up switching this in the python library. In fact, the library in its current state doesn't reference or use chat-templating-specific logic. I don't mind calling it `@huggingface/jinja`, but it should be clear to users what they are getting: something which is not a truly jinja-compatible templating library. The library is very limited in scope, and cannot (and probably shouldn't) support all functionality of a full templating engine.

For users who do want a feature-complete jinja-like templating engine, they can use [nunjucks](https://github.com/mozilla/nunjucks), however, that is not suitable for our needs since (as stated in [getting started guide](https://mozilla.github.io/nunjucks/getting-started.html)):

> nunjucks does not sandbox execution so it is not safe to run user-defined templates or inject user-defined content into template definitions. On the server, you can expose attack vectors for accessing sensitive data and remote code execution. On the client, you can expose cross-site scripting vulnerabilities even for precompiled templates (which can be mitigated with a strong [CSP](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy)). See [this issue](https://github.com/mozilla/nunjucks-docs/issues/17) for more information.

And all our templates are user-defined. 😅 nunjucks works by converting the template string to javascript code, then evaluating it as javascript code (see [this line of code](https://github.com/mozilla/nunjucks/blob/ea0d6d5396d39d9eed1b864febb36fbeca908f23/nunjucks/src/environment.js#L565)). In contrast, this library never evaluates anything with `eval`, `new Function`, or anything else like it. Instead, we create our [own runtime](https://github.com/huggingface/huggingface.js/blob/277163baf450bacf0a96750a5fe2a7d4269da7f0/packages/chat-templates/src/jinja/runtime.ts), and only expose very specific functionality to the user. See [here](https://github.com/huggingface/huggingface.js/blob/277163baf450bacf0a96750a5fe2a7d4269da7f0/packages/chat-templates/src/jinja/runtime.ts#L58-L77) for example, to see how we handle instance methods.

Naming it `@huggingface/jinja` will definitely help with SEO though 😆

**TLDR**: We could, but it might be misleading to users.

Comment (julien-c) [reply pnpm-workspace.yaml]: i'd call it `@huggingface/jinja` then if there's nothing specific to chat-templates. But i'll let you decide!

**Comment by xenova** [comment]
Great! 🚀 I'm just adding a bit more functionality first: mainly array slicing (needed by [llama](https://github.com/huggingface/transformers/blob/2098d343cc4b4b9d2aea84b3cf1eb5a1e610deff/src/transformers/models/llama/tokenization_llama.py#L439-L466)), and I'll add some more e2e tests.

Looking forward to the first release! 🔥

**Comment by xenova** [comment]
The library is now working for all chat templates in transformers, as well as [zephyr](https://huggingface.co/HuggingFaceH4/zephyr-7b-beta/blob/main/tokenizer_config.json#L34), which is more complicated.

Before launch, I'd like to add unit tests for invalid templates (and perhaps add more chat template examples; cc @Rocketknight1?).

Also, I think we can defer docs to a later date once more people have tested it and things are more finalized (hopefully no issues, but you never know 😉).

**Comment by julien-c** [comment]
IMO we could restrict the scope of Jinja features that can be used in transformers (and our other OS libraries)

i.e. The full power of Jinja expressivity is probably not needed, and it will be simpler to support a subset

**Comment by Rocketknight1** [comment]
Strongly agree - I want to keep the system open enough to support other use cases besides just `('system', 'assistant', 'user')` in the future, but many of Jinja's tools are not really relevant for chat templating at all, and we could definitely 'ban' them. We could also enforce other restrictions in Python (e.g. everything outside blocks is removed during preprocessing) that might make it much simpler to implement parsers.

**Comment by Rocketknight1** [comment]
Will take a look at the tests today!

**Comment by Rocketknight1** [comment]
Hi @xenova - the test suite looks great! The range of templates is good, and honestly if you're rendering the very complex llama one correctly then I'm 99% sure the implementation is good.

One question, since I'm not too good at reading JS test suites - are the templates tested with `add_generation_prompt` set and unset, to make sure the behaviour changes appropriately?

**Comment by xenova** [comment]
> are the templates tested with add_generation_prompt set and unset, to make sure the behaviour changes appropriately?

Sure, I added a test case for this (from [this page](https://huggingface.co/docs/transformers/main/en/chat_templating#how-do-i-use-chat-templates)). :) Ultimately, it doesn't test much more than those in templates.test.js (since that tests the core jinja functionality), but it's nice to have an end-to-end test case like that.

I'll also make sure that when integrating into transformers.js, the code (and output) will be identical to the python version.

**Comment by xenova** [comment]
@coyotte508 I've put together a simple [demo](https://huggingface.co/spaces/Xenova/jinja-testing) for the library:

https://github.com/huggingface/huggingface.js/assets/26504141/5c28d0e6-cd3f-4909-8cc5-617d1d05d677

Is it advised to publish the demo source code here too, or will that add unnecessary bloat? If the latter, I'll publish it to a separate repo.

**Comment by xenova** [comment]
> It can be done in a later PR if you want.

Yes let's do it in a follow-up PR! I'd like to get this published (at least a v0.0.1) to unblock chat templates in transformers.js 🔥 

Is there anything else needed to be done before publishing?

**Comment by xenova** [comment]
Great! So if I merge, now, will a new release be published? Or is that a separate thing to be done?

File: .github/workflows/chat-templates-publish.yml [hunk 3]
```diff
@@ -0,0 +1,51 @@
+name: Jinja - Version and Release
+
```
[the remark this answers is not in the document]
Comment (xenova) [reply .github/workflows/chat-templates-publish.yml]: Done 👍
